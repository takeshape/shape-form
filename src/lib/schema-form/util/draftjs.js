import {Map, List, Iterable, fromJS} from 'immutable';
import {convertToRaw, convertFromRaw, EditorState, ContentState} from 'draft-js';

function deserializeEditorState(data) {
  let contentState;
  if (typeof data === 'string') {
    contentState = ContentState.createFromText(data);
  } else {
    contentState = convertFromRaw(Iterable.isIterable(data) ? data.toJS() : data);
  }
  return EditorState.createWithContent(contentState);
}

function serializeEditorState(editorState) {
  return convertToRaw(editorState.getCurrentContent());
}

function serializerFactory(convert) {
  const transform = (schema, data, toPojo = false) => {
    if (data === null || data === undefined) {
      return data;
    }

    const type = schema.get('type');
    if (type === 'object') {
      if (schema.get('draftjs')) {
        return convert(data);
      }

      const fromImmutable = Iterable.isIterable(data);
      const properties = schema.get('properties');
      if (properties) {
        const result = {};
        properties.forEach((propSchema, name) => {
          const value = fromImmutable ? data.get(name) : data[name];
          if (value !== undefined) {
            result[name] = transform(propSchema, value, toPojo);
          }
        });

        return toPojo ? result : Map(result);
      }
      if (fromImmutable) {
        return toPojo ? data.toJS() : data;
      }
      return toPojo ? data : fromJS(data);
    }
    if (type === 'array') {
      const items = schema.get('items');
      if (items) {
        const mapped = data.map(item => transform(items, item, toPojo));
        if (Iterable.isIterable(data)) {
          return toPojo ? mapped.toArray() : mapped;
        }

        return toPojo ? mapped : List(mapped);
      }
    }

    return data;
  };

  return transform;
}

export const serializeDraftJs = serializerFactory(serializeEditorState);
export const deserializeDraftJs = serializerFactory(deserializeEditorState);

export function toPojo(obj, transform) {
  const convert = obj => {
    if (Iterable.isKeyed(obj)) {
      return obj
        .toSeq()
        .map(convert)
        .toObject();
    }

    if (Iterable.isIndexed(obj)) {
      return obj
        .toSeq()
        .map(convert)
        .toArray();
    }

    return transform ? transform(obj) : obj;
  };

  return convert(obj);
}
