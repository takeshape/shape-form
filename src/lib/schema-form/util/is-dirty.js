import {is} from 'immutable';
import {EditorState} from 'draft-js';

export default function isDirty(initialData, currentData) {
  if (currentData instanceof EditorState) {
    if (initialData) {
      return !is(initialData.getCurrentContent().getBlockMap(), currentData.getCurrentContent().getBlockMap());
    }
    return currentData.getCurrentContent().hasText();
  }
  if (initialData === undefined) {
    return currentData !== '' && currentData !== null && currentData !== undefined;
  }
  return !is(initialData, currentData);
}
