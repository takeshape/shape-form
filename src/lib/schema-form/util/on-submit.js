import {SUBMIT_FORM} from '../action-types';

export const onFormSubmit = formName => action =>
  action.type === `${SUBMIT_FORM}_SUCCESS` && action.meta.formName === formName;
