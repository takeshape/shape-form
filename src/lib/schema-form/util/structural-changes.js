import {Map, Range} from 'immutable';
import {pathToString, PROPERTIES} from '../paths';

function visitProps(schema, data, callback, path = []) {
  callback(schema, data, path);
  if (data === null || data === undefined) {
    return;
  }

  const type = schema.get('type');
  if (type === 'object') {
    const properties = schema.get('properties');
    if (properties) {
      properties.forEach((propSchema, name) => {
        const value = data.get(name);
        if (value !== undefined) {
          visitProps(propSchema, value, callback, path.concat(name));
        }
      });
    }
  } else if (type === 'array') {
    const itemSchema = schema.get('items');
    if (itemSchema) {
      return data.map((item, i) => visitProps(itemSchema, item, callback, path.concat(i)));
    }
  }
}

function initStructuralChangeProps(initialValue) {
  const size = initialValue.size;
  return Map({
    arrayKeys: Range(0, size).toList(),
    nextKey: size,
    originalSize: size
  });
}

export function initStructuralChanges(schema, initialData) {
  return Map().withMutations(result => {
    visitProps(schema, initialData, (schema, data, path) => {
      if (schema.get('type') === 'array') {
        result.setIn(path.concat(PROPERTIES), initStructuralChangeProps(data));
      }
    });
  });
}

function wasStructureModified(indicies, originalSize) {
  if (indicies.size !== originalSize) {
    return true;
  }
  for (let i = 0; i < indicies.size; i++) {
    if (indicies.get(i) !== i) {
      return true;
    }
  }
  return false;
}

export function flattenStructuralChanges(changes, path = [], result = []) {
  const props = changes.get(PROPERTIES);
  if (props) {
    const arrayKeys = props.get('arrayKeys');
    const originalSize = props.get('originalSize');

    if (wasStructureModified(arrayKeys, originalSize)) {
      result.push({path: pathToString(path), structure: arrayKeys.toArray()});
    }
  }
  changes.forEach((change, key) => {
    if (key !== PROPERTIES) {
      flattenStructuralChanges(change, path.concat(key), result);
    }
  });

  return result;
}
