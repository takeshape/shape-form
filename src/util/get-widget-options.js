import get from 'lodash/get';

export default function getWidgetOptions(schema, config) {
  let options = [];

  const configuredOptions = config.options;
  const sharedNameValueEnum = schema.enum;
  if (configuredOptions) {
    options = configuredOptions;
  } else if (sharedNameValueEnum) {
    options = sharedNameValueEnum.map(tzName => ({
      label: tzName,
      value: tzName
    }));
  } else {
    const oneOf = schema.type === 'array' ? get(schema, ['items', 'oneOf']) : schema.oneOf;
    if (oneOf) {
      options = oneOf.map(optionSchema => ({
        label: optionSchema.title,
        value: get(optionSchema, ['enum', 0])
      }));
    }
  }
  return options;
}
