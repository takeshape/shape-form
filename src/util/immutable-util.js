import merge from 'lodash/merge';
import set from 'lodash/set';
import update from 'lodash/update';

export const updateIn = update;
export const setIn = set;

export function removeIn(map, keyPath) {
  const lastIndex = keyPath.length - 1;
  return updateIn(map, keyPath.slice(0, lastIndex), parent => {
    if (typeof parent === 'object') {
      delete parent[keyPath[lastIndex]];
    }
    return parent;
  });
}

export function mergeIn(map, keyPath, value) {
  return updateIn(map, keyPath, current => ({...current, ...value}));
}

export function mergeInDeep(map, keyPath, value) {
  return updateIn(map, keyPath, current => merge(current, value));
}

export function removeIndex(array, i) {
  array.splice(i, 1);
  return array;
}
