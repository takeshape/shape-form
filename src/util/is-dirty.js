import isEqual from 'lodash/isEqual';

export default function isDirty(initialData, currentData) {
  if (initialData === undefined) {
    return currentData !== '' && currentData !== null && currentData !== undefined;
  }
  return !isEqual(initialData, currentData);
}
