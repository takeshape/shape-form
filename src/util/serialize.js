function serializerFactory() {
  const transform = (schema, data) => {
    if (data === null || data === undefined) {
      return data;
    }

    const type = schema.type;
    if (type === 'object') {
      if (schema.draftjs) {
        return null;
      }

      const properties = schema.properties;
      if (properties) {
        const result = {};
        for (const name of Object.keys(properties)) {
          const propSchema = properties[name];
          const value = data[name];
          if (value !== undefined) {
            result[name] = transform(propSchema, value);
          }
        }
        return result;
      }
      return data;
    }
    if (type === 'array') {
      const items = schema.items;
      if (items) {
        return data.map(item => transform(items, item));
      }
    }

    return data;
  };

  return transform;
}

export const serialize = serializerFactory();
