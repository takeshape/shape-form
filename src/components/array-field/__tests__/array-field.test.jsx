import produce from 'immer';

jest.mock('../../schema-field', () => 'SchemaField');
import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';

const renderer = new ReactShallowRenderer();
import {shallow} from 'enzyme';
import ArrayField from '../array-field';
import Fab from '@material-ui/core/Fab';
import {setIn} from '../../../util/immutable-util';

const formName = 'test-form';
const schema = {
  type: 'array',
  items: {
    type: 'object',
    properties: {
      email: {type: 'string'},
      firstName: {type: 'string'},
      lastName: {type: 'string'}
    }
  }
};

const ui = {
  arrayKeys: [0, 1, 2],
  collapsed: new Set()
};

function createProps(props) {
  return {
    schema,
    formName,
    ui,
    path: 'users',
    config: {},
    context: {},
    widgets: {},
    isRequired: false,

    length: 3,
    fieldAction: jest.fn(),
    addArrayItem: jest.fn(),
    removeArrayItem: jest.fn(),
    toggleArrayItem: jest.fn(),
    expandAllArrayItems: jest.fn(),
    collapseAllArrayItems: jest.fn(),
    initializeArray: jest.fn(),
    blurField: jest.fn(),
    locale: 'pt-br',
    ...props
  };
}

function mockEvent() {
  return {
    preventDefault: jest.fn()
  };
}

describe('ArrayField', () => {
  it('renders', () => {
    const props = createProps();
    const tree = renderer.render(<ArrayField {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders when disabled', () => {
    const props = createProps();
    const tree = renderer.render(<ArrayField {...props} disabled />);
    expect(tree).toMatchSnapshot();
  });

  it('renders empty', () => {
    const props = createProps({
      length: 0
    });

    const tree = renderer.render(<ArrayField {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders single', () => {
    const props = createProps({
      schema: {...schema, minItems: 1, maxItems: 1},
      ui: {...ui, arrayKeys: [0], collapsed: new Set()},
      path: 'user',
      length: 1
    });

    const tree = renderer.render(<ArrayField {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders placeholder', () => {
    const props = createProps({
      ui: {...ui, placeholder: {index: 1}}
    });

    const tree = renderer.render(<ArrayField {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders with dragAndDrop', () => {
    const props = createProps({
      ui: {...ui, placeholder: {index: 1}},
      config: {dragAndDrop: true}
    });

    const tree = renderer.render(<ArrayField {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders with a custom wrapper', () => {
    const Unstyled = ({children}) => <div>{children}</div>;// eslint-disable-line
    const props = createProps({
      widgets: {unstyled: Unstyled},
      config: {wrapper: 'unstyled'}
    });

    const tree = renderer.render(<ArrayField {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('calls initializeArray on mount', () => {
    const props = createProps();

    shallow(<ArrayField {...props} />);
    expect(props.initializeArray).toHaveBeenCalledTimes(1);
    expect(props.initializeArray).toHaveBeenCalledWith(props.formName, props.path);
  });

  it('calls initializeArray on update when the path changes', () => {
    const props = createProps();

    const wrapper = shallow(<ArrayField {...props} />);
    expect(props.initializeArray).toHaveBeenCalledTimes(1);
    expect(props.initializeArray).toHaveBeenLastCalledWith(props.formName, props.path);

    wrapper.setProps({error: 'OH NO'});
    expect(props.initializeArray).toHaveBeenCalledTimes(1);

    wrapper.setProps({path: 'new'});
    expect(props.initializeArray).toHaveBeenCalledTimes(2);
    expect(props.initializeArray).toHaveBeenLastCalledWith(props.formName, 'new');
  });

  it('handleAddItem', () => {
    const props = createProps();

    const wrapper = shallow(<ArrayField {...props} />);
    const event = mockEvent();

    wrapper.find(Fab).simulate('click', event);
    expect(event.preventDefault).toHaveBeenCalledTimes(1);
    expect(props.addArrayItem).toHaveBeenCalledTimes(1);
    expect(props.addArrayItem).toHaveBeenCalledWith({});
    expect(props.blurField).toHaveBeenCalledTimes(1);
  });

  it('handleAddItem with default value', () => {
    const defaultValue = {
      email: 'samy@baguette.fr',
      firstName: 'Samy',
      lastName: 'PessÃ©'
    };
    const props = produce(createProps(), draft => setIn(draft, ['schema', 'items', 'default'], defaultValue));

    const wrapper = shallow(<ArrayField {...props} />);
    const event = mockEvent();
    wrapper.find(Fab).simulate('click', event);
    expect(event.preventDefault).toHaveBeenCalledTimes(1);
    expect(props.addArrayItem).toHaveBeenCalledTimes(1);
    expect(props.addArrayItem).toHaveBeenCalledWith(defaultValue);
    expect(props.blurField).toHaveBeenCalledTimes(1);
  });
});
