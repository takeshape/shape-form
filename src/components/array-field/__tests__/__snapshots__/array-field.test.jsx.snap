// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayField renders 1`] = `
<ForwardRef(WithStyles)
  actions={
    <span>
      <ForwardRef(WithStyles)
        onClick={[MockFunction]}
        size="small"
      >
        <Memo(UnfoldLessIcon) />
      </ForwardRef(WithStyles)>
    </span>
  }
  addArrayItem={[MockFunction]}
  blurField={[MockFunction]}
  collapseAllArrayItems={[MockFunction]}
  config={Object {}}
  context={Object {}}
  expandAllArrayItems={[MockFunction]}
  fieldAction={[MockFunction]}
  formName="test-form"
  initializeArray={[MockFunction]}
  isRequired={false}
  length={3}
  locale="pt-br"
  path="users"
  removeArrayItem={[MockFunction]}
  schema={
    Object {
      "items": Object {
        "properties": Object {
          "email": Object {
            "type": "string",
          },
          "firstName": Object {
            "type": "string",
          },
          "lastName": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    }
  }
  toggleArrayItem={[MockFunction]}
  ui={
    Object {
      "arrayKeys": Array [
        0,
        1,
        2,
      ],
      "collapsed": Set {},
    }
  }
  widgets={Object {}}
>
  <ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      arrayKey={0}
      config={Object {}}
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={0}
      index={0}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
    <ForwardRef(WithStyles)
      arrayKey={1}
      config={Object {}}
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={1}
      index={1}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
    <ForwardRef(WithStyles)
      arrayKey={2}
      config={Object {}}
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={2}
      index={2}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
  </ForwardRef(WithStyles)>
  <ForwardRef(WithStyles)
    color="secondary"
    onClick={[Function]}
    size="small"
  >
    <Memo(AddIcon) />
  </ForwardRef(WithStyles)>
</ForwardRef(WithStyles)>
`;

exports[`ArrayField renders empty 1`] = `
<ForwardRef(WithStyles)
  addArrayItem={[MockFunction]}
  blurField={[MockFunction]}
  collapseAllArrayItems={[MockFunction]}
  config={Object {}}
  context={Object {}}
  expandAllArrayItems={[MockFunction]}
  fieldAction={[MockFunction]}
  formName="test-form"
  initializeArray={[MockFunction]}
  isRequired={false}
  length={0}
  locale="pt-br"
  path="users"
  removeArrayItem={[MockFunction]}
  schema={
    Object {
      "items": Object {
        "properties": Object {
          "email": Object {
            "type": "string",
          },
          "firstName": Object {
            "type": "string",
          },
          "lastName": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    }
  }
  toggleArrayItem={[MockFunction]}
  ui={
    Object {
      "arrayKeys": Array [
        0,
        1,
        2,
      ],
      "collapsed": Set {},
    }
  }
  widgets={Object {}}
>
  <ForwardRef(WithStyles) />
  <ForwardRef(WithStyles)
    color="secondary"
    onClick={[Function]}
    size="small"
  >
    <Memo(AddIcon) />
  </ForwardRef(WithStyles)>
</ForwardRef(WithStyles)>
`;

exports[`ArrayField renders placeholder 1`] = `
<ForwardRef(WithStyles)
  actions={
    <span>
      <ForwardRef(WithStyles)
        onClick={[MockFunction]}
        size="small"
      >
        <Memo(UnfoldLessIcon) />
      </ForwardRef(WithStyles)>
    </span>
  }
  addArrayItem={[MockFunction]}
  blurField={[MockFunction]}
  collapseAllArrayItems={[MockFunction]}
  config={Object {}}
  context={Object {}}
  expandAllArrayItems={[MockFunction]}
  fieldAction={[MockFunction]}
  formName="test-form"
  initializeArray={[MockFunction]}
  isRequired={false}
  length={3}
  locale="pt-br"
  path="users"
  removeArrayItem={[MockFunction]}
  schema={
    Object {
      "items": Object {
        "properties": Object {
          "email": Object {
            "type": "string",
          },
          "firstName": Object {
            "type": "string",
          },
          "lastName": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    }
  }
  toggleArrayItem={[MockFunction]}
  ui={
    Object {
      "arrayKeys": Array [
        0,
        1,
        2,
      ],
      "collapsed": Set {},
      "placeholder": Object {
        "index": 1,
      },
    }
  }
  widgets={Object {}}
>
  <ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      arrayKey={0}
      config={Object {}}
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={0}
      index={0}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
    <ForwardRef(WithStyles)
      arrayKey={666}
      config={Object {}}
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={1}
      index={1}
      isCollapsed={false}
      isPlaceholder={true}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
    <ForwardRef(WithStyles)
      arrayKey={1}
      config={Object {}}
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={2}
      index={1}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
    <ForwardRef(WithStyles)
      arrayKey={2}
      config={Object {}}
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={3}
      index={2}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
  </ForwardRef(WithStyles)>
  <ForwardRef(WithStyles)
    color="secondary"
    onClick={[Function]}
    size="small"
  >
    <Memo(AddIcon) />
  </ForwardRef(WithStyles)>
</ForwardRef(WithStyles)>
`;

exports[`ArrayField renders single 1`] = `
<SchemaField
  config={Object {}}
  context={Object {}}
  formName="test-form"
  locale="pt-br"
  path="user[0]"
  schema={
    Object {
      "properties": Object {
        "email": Object {
          "type": "string",
        },
        "firstName": Object {
          "type": "string",
        },
        "lastName": Object {
          "type": "string",
        },
      },
      "type": "object",
    }
  }
  widgets={Object {}}
/>
`;

exports[`ArrayField renders when disabled 1`] = `
<ForwardRef(WithStyles)
  actions={
    <span>
      <ForwardRef(WithStyles)
        onClick={[MockFunction]}
        size="small"
      >
        <Memo(UnfoldLessIcon) />
      </ForwardRef(WithStyles)>
    </span>
  }
  addArrayItem={[MockFunction]}
  blurField={[MockFunction]}
  collapseAllArrayItems={[MockFunction]}
  config={Object {}}
  context={Object {}}
  disabled={true}
  expandAllArrayItems={[MockFunction]}
  fieldAction={[MockFunction]}
  formName="test-form"
  initializeArray={[MockFunction]}
  isRequired={false}
  length={3}
  locale="pt-br"
  path="users"
  removeArrayItem={[MockFunction]}
  schema={
    Object {
      "items": Object {
        "properties": Object {
          "email": Object {
            "type": "string",
          },
          "firstName": Object {
            "type": "string",
          },
          "lastName": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    }
  }
  toggleArrayItem={[MockFunction]}
  ui={
    Object {
      "arrayKeys": Array [
        0,
        1,
        2,
      ],
      "collapsed": Set {},
    }
  }
  widgets={Object {}}
>
  <ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      arrayKey={0}
      config={Object {}}
      context={Object {}}
      disabled={true}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={0}
      index={0}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
    <ForwardRef(WithStyles)
      arrayKey={1}
      config={Object {}}
      context={Object {}}
      disabled={true}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={1}
      index={1}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
    <ForwardRef(WithStyles)
      arrayKey={2}
      config={Object {}}
      context={Object {}}
      disabled={true}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={2}
      index={2}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
  </ForwardRef(WithStyles)>
</ForwardRef(WithStyles)>
`;

exports[`ArrayField renders with a custom wrapper 1`] = `
<Unstyled
  addArrayItem={[MockFunction]}
  blurField={[MockFunction]}
  collapseAllArrayItems={[MockFunction]}
  config={
    Object {
      "wrapper": "unstyled",
    }
  }
  context={Object {}}
  expandAllArrayItems={[MockFunction]}
  fieldAction={[MockFunction]}
  formName="test-form"
  initializeArray={[MockFunction]}
  isRequired={false}
  length={3}
  locale="pt-br"
  onCollapseAll={[MockFunction]}
  onExpandAll={[MockFunction]}
  path="users"
  removeArrayItem={[MockFunction]}
  schema={
    Object {
      "items": Object {
        "properties": Object {
          "email": Object {
            "type": "string",
          },
          "firstName": Object {
            "type": "string",
          },
          "lastName": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    }
  }
  toggleArrayItem={[MockFunction]}
  ui={
    Object {
      "arrayKeys": Array [
        0,
        1,
        2,
      ],
      "collapsed": Set {},
    }
  }
  widgets={
    Object {
      "unstyled": [Function],
    }
  }
>
  <ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      arrayKey={0}
      config={
        Object {
          "wrapper": "unstyled",
        }
      }
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={0}
      index={0}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={
        Object {
          "unstyled": [Function],
        }
      }
    />
    <ForwardRef(WithStyles)
      arrayKey={1}
      config={
        Object {
          "wrapper": "unstyled",
        }
      }
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={1}
      index={1}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={
        Object {
          "unstyled": [Function],
        }
      }
    />
    <ForwardRef(WithStyles)
      arrayKey={2}
      config={
        Object {
          "wrapper": "unstyled",
        }
      }
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={2}
      index={2}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={
        Object {
          "unstyled": [Function],
        }
      }
    />
  </ForwardRef(WithStyles)>
  <ForwardRef(WithStyles)
    color="secondary"
    onClick={[Function]}
    size="small"
  >
    <Memo(AddIcon) />
  </ForwardRef(WithStyles)>
</Unstyled>
`;

exports[`ArrayField renders with dragAndDrop 1`] = `
<ForwardRef(WithStyles)
  actions={
    <span>
      <ForwardRef(WithStyles)
        onClick={[MockFunction]}
        size="small"
      >
        <Memo(UnfoldLessIcon) />
      </ForwardRef(WithStyles)>
    </span>
  }
  addArrayItem={[MockFunction]}
  blurField={[MockFunction]}
  collapseAllArrayItems={[MockFunction]}
  config={
    Object {
      "dragAndDrop": true,
    }
  }
  context={Object {}}
  expandAllArrayItems={[MockFunction]}
  fieldAction={[MockFunction]}
  formName="test-form"
  initializeArray={[MockFunction]}
  isRequired={false}
  length={3}
  locale="pt-br"
  path="users"
  removeArrayItem={[MockFunction]}
  schema={
    Object {
      "items": Object {
        "properties": Object {
          "email": Object {
            "type": "string",
          },
          "firstName": Object {
            "type": "string",
          },
          "lastName": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    }
  }
  toggleArrayItem={[MockFunction]}
  ui={
    Object {
      "arrayKeys": Array [
        0,
        1,
        2,
      ],
      "collapsed": Set {},
      "placeholder": Object {
        "index": 1,
      },
    }
  }
  widgets={Object {}}
>
  <DropTarget(ArrayFieldDropTarget)
    count={3}
    fieldAction={[MockFunction]}
    path="users"
    placeholderIndex={1}
  >
    <ForwardRef(WithStyles)
      arrayKey={0}
      config={
        Object {
          "dragAndDrop": true,
        }
      }
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={0}
      index={0}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
    <ForwardRef(WithStyles)
      arrayKey={666}
      config={
        Object {
          "dragAndDrop": true,
        }
      }
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={1}
      index={1}
      isCollapsed={false}
      isPlaceholder={true}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
    <ForwardRef(WithStyles)
      arrayKey={1}
      config={
        Object {
          "dragAndDrop": true,
        }
      }
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={2}
      index={1}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
    <ForwardRef(WithStyles)
      arrayKey={2}
      config={
        Object {
          "dragAndDrop": true,
        }
      }
      context={Object {}}
      fieldAction={[MockFunction]}
      formName="test-form"
      hoverIndex={3}
      index={2}
      isCollapsed={false}
      isPlaceholder={false}
      locale="pt-br"
      onBlur={[MockFunction]}
      path="users"
      removeItem={[MockFunction]}
      schema={
        Object {
          "properties": Object {
            "email": Object {
              "type": "string",
            },
            "firstName": Object {
              "type": "string",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "type": "object",
        }
      }
      toggleItem={[MockFunction]}
      widgets={Object {}}
    />
  </DropTarget(ArrayFieldDropTarget)>
  <ForwardRef(WithStyles)
    color="secondary"
    onClick={[Function]}
    size="small"
  >
    <Memo(AddIcon) />
  </ForwardRef(WithStyles)>
</ForwardRef(WithStyles)>
`;
